### PR TITLE
Add document visibility scope

### PR DESCRIPTION
**Pull request by osallou**
Only space owners can upload/edit/delete documents
On edit document has a visibility scope to give download rights
Space managers have all rights

* public: anyone
* members: space members only
* user: for a specific user
* client: all user for this client
* private: space managers only

Set a flaggable feature, in plan 1 in example conf.yaml file

### CODE REVIEW

File: Modules/documents/Controller/DocumentslistController.php [hunk 1]
```diff
@@ -105,6 +260,11 @@ public function openAction($id_space, $id) {
 
     public function deleteAction($id_space, $id) {
         $this->checkAuthorizationMenuSpace("documents", $id_space, $_SESSION["id_user"]);
+        $modelCoreConfig = new CoreConfig();
+        $editRole = $modelCoreConfig->getParamSpace("documentsEdit", $id_space, CoreSpace::$MANAGER);
+        if($this->role < $editRole) {
+            throw new PfmAuthException('not enough priviledges');
```
Comment (gaetanHerve) [inline Modules/documents/Controller/DocumentslistController.php:266]: just a typo: "privileges"

File: Modules/core/Model/CoreSpace.php [hunk 2]
```diff
@@ -22,6 +22,7 @@ class CorePlan {
      */
```
Comment (gaetanHerve) [inline Modules/core/Model/CoreSpace.php:22]: Shouldn't this file be named CorePlan.php ?

File: Modules/documents/Controller/DocumentslistController.php [hunk 3]
```diff
@@ -22,38 +27,114 @@ class DocumentslistController extends DocumentsController {
...
+            $visibility = 'Private';
+            switch ($data[$i]['visibility']) {
+                case Document::$VISIBILITY_PUBLIC:
+                    $visibility = 'Public';
+                    break;
+                case Document::$VISIBILITY_MEMBERS:
+                    $visibility = 'Members';
+                    break;
+                case Document::$VISIBILITY_USER:
+                    $visibility = 'User'.$data[$i]['id_ref'];
```
Comment (gaetanHerve) [inline Modules/documents/Controller/DocumentslistController.php:91]: would be more convenient to display user name or login

File: Modules/documents/Controller/DocumentslistController.php [hunk 4]
```diff
@@ -22,38 +27,114 @@ class DocumentslistController extends DocumentsController {
...
+                    break;
+                case Document::$VISIBILITY_CLIENT:
+                    $visibility = 'Client '.$data[$i]['id_ref'];
```
Comment (gaetanHerve) [inline Modules/documents/Controller/DocumentslistController.php:94]: Same comment than above

File: Modules/documents/Controller/DocumentslistController.php [hunk 5]
```diff
@@ -62,6 +143,30 @@ public function editAction($id_space, $id) {
...
+                $data['visibility'] ?? Document::$VISIBILITY_PRIVATE
+            );
+            $mu = new CoreUser();
+            $users = $mu->getSpaceActiveUsersForSelect($id_space, "name");
+            $mc = new ClClient();
+            $clients = $mc->getForList($id_space);
+            $clients['names'] = array_merge([''], $clients['names']);
+            $clients['ids'] = array_merge([0], $clients['ids']);
+            $form->addSelect('id_ref_user', CoreTranslator::User($lang), $users['names'], $users['ids'], $data['id_ref'] ?? 0);
+            $form->addSelect('id_ref_client', ClientsTranslator::Client($lang), $clients['names'], $clients['ids'], $data['id_ref'] ?? 0);
```
Comment (gaetanHerve) [inline Modules/documents/Controller/DocumentslistController.php:164]: An ideal would be to display only one (or none) of these two inputs based on visibility choice. It is possible to use DynamicForms.js to do that dynamically (but need to add a new method/option for displaying / hiding inputs)

**Review by gaetanHerve — COMMENTED** [review]
I add a few comments. Tests ok, but I had to bypass plans relative conditions (maybe wrong configuration on my environment)

**Comment by osallou** [comment]
@gaetanHerve I made some update to display docs in a tree way

doc "names" can use virtual dirs "public/toto" for example and this will impact display

better to organize a bit the docs

File: Modules/documents/Model/Document.php [hunk 6]
```diff
@@ -49,6 +57,28 @@ public function edit($id, $id_space, $title, $id_user){
...
         $this->runRequest($sql, array($title, $id_user, $id, $id_space));
     }
+
+    public function setVisibility($id_space, $id, $visibility, $id_ref) {
+        $sql = "UPDATE dc_documents SET visibility=?, id_ref=? WHERE id=? AND id_space=?";
+        $this->runRequest($sql, array($visibility, $id_ref, $id, $id_space));   
+    }
+
+    public function getPublicDocs($id_space){
+        $sql = "SELECT dc_documents.*, core_users.login as user FROM dc_documents INNER JOIN core_users on core_users.id=dc_documents.id_user WHERE id_space=? AND visibility=?";
```
Comment (gaetanHerve) [inline Modules/documents/Model/Document.php:67]: returns a SQL error 'id_space' is ambiguous. should be WHERE dc_documents.id_space=? AND visibility=?";

Comment (osallou) [reply Modules/documents/Model/Document.php:67]: yes

**Review by gaetanHerve — APPROVED** [review]
Tested and approved :)

**Pull request merged**